File: 2320-find-all-k-distant-indices-in-an-array/find-all-k-distant-indices-in-an-array.java
class Solution {
    public List<Integer> findKDistantIndices(int[] nums, int key, int k) {
        List<Integer> result = new ArrayList<>();
        int n = nums.length, size = 0, lastIndex = -1;
        for (int i = 0; i < n; i++) {
            if (key == nums[i]) {
                int start = Math.max(0, Math.max(i-k, lastIndex + 1));
                int end = Math.min(n-1, i + k);
                for (int j = start; j <= end; j++) {
                    result.add(j);
                    lastIndex = j;
                }
            }
        }

        return result;
    }
}

File: 2320-find-all-k-distant-indices-in-an-array/find-all-k-distant-indices-in-an-array.cpp
class Solution {
public:
    vector<int> findKDistantIndices(vector<int>& nums, int key, int k) {
        int n = nums.size();
        int lastIndex = -1;
        vector<int> result;
        for (int i = 0; i < n; i++) {
            if (nums[i] == key) {
                int start = std::max(0, std::max(i - k, lastIndex + 1));
                int end = std::min(n-1, i + k);
                for (int j = start; j <= end; j++) {
                    result.push_back(j);
                    lastIndex = j;
                }
            }
        }

        return result;
    }
};